go/gossip/storage.go

```diff
@@ -0,0 +1,339 @@
+
+        CREATE TABLE IF NOT EXISTS scts (
+                sct_id  INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+                sct     BYTES NOT NULL UNIQUE
+        );
+
+        CREATE TABLE IF NOT EXISTS chains (
+                chain_id    INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+                chain       STRING NOT NULL UNIQUE
+        );
+
+        CREATE TABLE IF NOT EXISTS sct_feedback (
+                chain_id    INTEGER NOT NULL REFERENCES chains(chain_id),
+                sct_id      INTEGER NOT NULL REFERENCES scts(sct_id),
+                PRIMARY KEY (chain_id, sct_id)
+
+        );`
+
+const insertChain = `INSERT INTO chains(chain) VALUES ($1);`
+const insertSCT = `INSERT INTO scts(sct) VALUES ($1);`
+const insertSCTFeedback = `INSERT INTO sct_feedback(chain_id, sct_id) VALUES ($1, $2);`
+const insertSTHPollination = `INSERT INTO sths(version, tree_size, timestamp, root_hash, signature, log_id) VALUES($1, $2, $3, $4, $5, $6);`
+
+const selectChainID = `SELECT chain_id FROM chains WHERE chain = $1;`
+
+// Selects at most $2 rows from the sths table whose timestamp is newer than $1.
+const selectRandomRecentPollination = `SELECT version, tree_size, timestamp, root_hash, signature, log_id FROM sths 
+                                          WHERE timestamp >= $1 ORDER BY random() LIMIT $2;`
+const selectSCTID = `SELECT sct_id FROM scts WHERE sct = $1;`
+
+const selectNumSCTs = `SELECT COUNT(*) FROM scts;`
+const selectNumChains = `SELECT COUNT(*) FROM chains;`
+const selectNumFeedback = `SELECT COUNT(*) FROM sct_feedback;`
+const selectNumSTHs = `SELECT COUNT(*) FROM sths;`
+const selectFeedback = `SELECT COUNT(*) FROM sct_feedback WHERE chain_id = $1 AND sct_id = $2;`
+const selectSTH = `SELECT COUNT(*) FROM sths WHERE version = $1 AND tree_size = $2 AND timestamp = $3 AND root_hash = $4 AND signature = $5 AND log_id = $6;`
+
+// Storage provides an SQLite3-backed method for persisting gossip data
+type Storage struct {
+	db                            *sql.DB
+	insertChain                   *sql.Stmt
+	insertSCT                     *sql.Stmt
+	insertSCTFeedback             *sql.Stmt
+	insertSTHPollination          *sql.Stmt
+	selectChainID                 *sql.Stmt
+	selectRandomRecentPollination *sql.Stmt
+	selectSCTID                   *sql.Stmt
+
+	selectNumChains   *sql.Stmt
+	selectNumFeedback *sql.Stmt
+	selectNumSCTs     *sql.Stmt
+	selectNumSTHs     *sql.Stmt
+
+	selectFeedback *sql.Stmt
+	selectSTH      *sql.Stmt
+}
+
+// Open opens the underlying persistent data store.
+// Should be called before attempting to use any of the store or search methods.
+func (s *Storage) Open() error {
+	var err error
+	if s.db != nil {
+		return errors.New("attempting to call Open() on an already Open()'d Storage")
+	}
+	s.db, err = sql.Open("sqlite3", *dbName)
+	if err != nil {
+		return err
+	}
+	if _, err := s.db.Exec(schema); err != nil {
+		return err
+	}
+	if s.insertChain, err = s.db.Prepare(insertChain); err != nil {
```

Comment:
One obvious possibility...

Most of these SQL strings could be synthesized on the fly (e.g "SELECT COUNT(*) FROM <table>"), you could keep a map from strings to prepared statements and prepare them on the fly.